File: pmod_dac.py
#   Copyright (c) 2016, Xilinx, Inc.
#   All rights reserved.
# 
#   Redistribution and use in source and binary forms, with or without 
#   modification, are permitted provided that the following conditions are met:
#
#   1.  Redistributions of source code must retain the above copyright notice, 
#       this list of conditions and the following disclaimer.
#
#   2.  Redistributions in binary form must reproduce the above copyright 
#       notice, this list of conditions and the following disclaimer in the 
#       documentation and/or other materials provided with the distribution.
#
#   3.  Neither the name of the copyright holder nor the names of its 
#       contributors may be used to endorse or promote products derived from 
#       this software without specific prior written permission.
#
#   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
#   THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
#   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
#   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
#   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
#   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
#   OR BUSINESS INTERRUPTION). HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
#   WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
#   OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
#   ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


"""
 * Passed parameters in MAILBOX_WRITE_CMD
 * bits 31:20 => delay in microsecond if wave generation mode is selected
 * bits 31:20 => 12-bit value to be output if fixed value mode is selected
 * bits 19:16 => 4-bit value indicates channels to be used
 * bits 15:8 => number of cycles to be produced in wave generation mode, 
 *              if it is set to 0, then generate waveforms forever
 * bit 7 => not used
 * bit 6 => random wave generation - pattern is written in MAILBOX_DATA area
 *          with MAILBOX_DATA(0) having number of samples per cycle, followed 
 *          by 12-bit values in each word
 * bit 5 => Sine wave generation mode
 * bit 4 => Triangle waveform generation mode
 * bit 3 => Sawtooth generation mode
 * bit 2 => Square wave generation mode
 * bit 1 => Fixed Value mode
 * bit 0 => 1 command issued, 0 command completed
"""


from . import Pmod
import struct
import csv
import time

PMOD_DAC_PROGRAM = "pmod_dac.bin"
FIXEDGEN = 0x3


class Pmod_DAC(object):
    """This class controls a Digital to Analog Converter Pmod.
    
    The Pmod DA4 (PB 200-245) is an 8 channel 12-bit digital-to-analog 
    converter run via AD5628.

    Attributes
    ----------
    microblaze : Pmod
        Microblaze processor instance used by this module.
        
    """

    def __init__(self, mb_info):
        """Return a new instance of a DAC object.
    
        Note
        ----
        The floating point number to be written should be in the range 
        of [0.00, 2.50]. 
        
        Parameters
        ----------
        mb_info : dict
            A dictionary storing Microblaze information, such as the
            IP name and the reset name.
            
        """
        self.microblaze = Pmod(mb_info, PMOD_DAC_PROGRAM)

    def write_fixed_value(self, value):
        """
        Write a floating point number onto the DAC Pmod.
        """
        if not 0.00 <= value <= 2.5:
            raise ValueError("Requested value not in range [0.00, 2.50].")

        # Calculate the voltage value and write to DAC
        int_val = int(value / 0.0006105)
        cmd = (int_val << 20) | FIXEDGEN
        self.microblaze.write_blocking_command(cmd)

    def write_arbitrary(self):
        wave = []
        convertedWave = []
        
        print('reading in wave.csv')
        with open('wave.csv') as csv_file:
            csv_reader = csv.reader(csv_file, delimiter=',')
            for row in csv_reader:
                wave.append(row[0])
        
        # only 253 adddresses each one is 32 bits
        if len(wave) > 253 * 2:
            print('wave.csv too long')
            return
        
        if max(wave) > 2.5:
            print('max value over 2.5V')
            return
        
        # encode voltage values into 32 bit integers using resolution/step size of 4096 and max value 2.5V
        # each 32 bit integer will contain two 12 bit values (2^12 = 4096) and padded in MSB with zeros
        for i in range(0, len(wave), 1):
            integer = 0
            
            # handle odd length
            if i == len(wave) - 1:
                value1, value2 = wave[i], 0
            else:    
                value1, value2 = wave[i], wave[i + 1]

            # convert to integer between 0 and 4096
            value1, value2 = int(value1 * 4096 / 2.5), int(value2 * 4096 / 2.5)

            value1 = value1 << 12
            integer += value1 + value2
            convertedWave.append(integer)
        
        print('writing to microblaze mailbox')

        # write to mailbox
        self.microblaze.write_mailbox(0, convertedWave)

        # arbitrary wait to make sure it done writing (don't know if write_mailbox is blocking)
        time.sleep(2)

        print('running arbitrary wave generator')
        
        cmd = 8 + 1 # mode=3 and lsb = 1 means run cmd
        self.microblaze.write_blocking_command(cmd)



    def write_sawtooth(self, cycles, delay):
        """
        cycles is the number of cycles of the wave we want to output
        before stopping the signal. If cycles is 0 the wave is generated forever

        the delay is the number of microseconds to wait between 
        writing each step of the wave
        """
        delay = (delay & 0xFFF) << 20
        cycles = (cycles & 0xFF) << 8
        mode = 8 
        cmd = cycles + delay + mode + 1
        self.microblaze.write_blocking_command(cmd)
    
    def write_square_wave(self, cycles, delay):
        """
        cycles is the number of cycles of the wave we want to output
        before stopping the signal. If cycles is 0 the wave is generated forever

        the delay is the number of microseconds to wait between 
        writing each step of the wave
        """
        delay = (delay & 0xFFF) << 20
        cycles = (cycles & 0xFF) << 8
        mode = 4 
        cmd = cycles + delay + mode + 1
        self.microblaze.write_blocking_command(cmd)
    
        def write_triangle_wave(self, cycles, delay):
        """
        cycles is the number of cycles of the wave we want to output
        before stopping the signal. If cycles is 0 the wave is generated forever

        the delay is the number of microseconds to wait between 
        writing each step of the wave
        """
        delay = (delay & 0xFFF) << 20
        cycles = (cycles & 0xFF) << 8
        mode = 16 
        cmd = cycles + delay + mode + 1
        self.microblaze.write_blocking_command(cmd)
    
    def write_sine_wave(self, cycles, delay):
        """
        cycles is the number of cycles of the wave we want to output
        before stopping the signal. If cycles is 0 the wave is generated forever

        the delay is the number of microseconds to wait between 
        writing each step of the wave
        """
        delay = (delay & 0xFFF) << 20
        cycles = (cycles & 0xFF) << 8
        mode = 32 
        cmd = cycles + delay + mode + 1
        self.microblaze.write_blocking_command(cmd)
    
    def write_sawtooth(self, cycles, delay):
        """
        cycles is the number of cycles of the wave we want to output
        before stopping the signal. If cycles is 0 the wave is generated forever

        the delay is the number of microseconds to wait between 
        writing each step of the wave
        """
        delay = (delay & 0xFFF) << 20
        cycles = (cycles & 0xFF) << 8
        mode = 8 # sawtooth mode
        cmd = cycles + delay + mode + 1
        self.microblaze.write_blocking_command(cmd)
    

File: workspace/pmod_dac/src/pmod_dac.c
#include "xparameters.h"
#include "circular_buffer.h"
#include "spi.h"
#include "timer.h"
#include "xparameters.h"
#include "xil_types.h"
#include "xil_io.h"

#define XPAR_PMOD_IO_SWITCH_BASEADDR XPAR_MB_1_MB1_SWITCH_S00_AXI_BASEADDR
#define MAILBOX_DATA(x) (*(volatile unsigned *)(0x0000FF00+((x)*4)))

static spi device;
/*
 * Passed parameters in MAILBOX_WRITE_CMD
 * bits 31:20 => delay in microsecond if wave generation mode is selected
 * bits 31:20 => 12-bit value to be output if fixed value mode is selected
 * bits 19:16 => 4-bit value indicates channels to be used
 * bits 15:8 => number of cycles to be produced in wave generation mode, 
 *              if it is set to 0, then generate waveforms forever
 * bit 0 => 1 command issued, 0 command completed
 *
 * Channels table
 * bits [19:16]
 * 0000 => Channel A
 * 0001 => Channel B
 * 0010 => Channel C
 * 0011 => Channel D
 * 0100 => Channel E
 * 0101 => Channel F
 * 0110 => Channel G
 * 1111 => All channels
 */

#define BUFFER_SIZE     6

typedef u8 DataBuffer[BUFFER_SIZE];

u8 WriteBuffer[BUFFER_SIZE];

void RefOn(void) {
    // Turn ON internal reference voltage
    WriteBuffer[3]=0x01;
    // This byte must be 0
    WriteBuffer[2]=0x00;
    // channel number (0) | least significant 4-bits data
    WriteBuffer[1]=0x00;
    // write to and update requested channel
    WriteBuffer[0]=0x08;
    spi_transfer(device, (char*)WriteBuffer, NULL, 4);
}

void RefOFF(void) {
    // Turn OFF internal reference voltage
    WriteBuffer[3]=0x00;
    // This byte must be 0
    WriteBuffer[2]=0x00;
    // channel number (0) | least significant 4-bits data
    WriteBuffer[1]=0x00;
    // write to and update requested channel
    WriteBuffer[0]=0x08;
    spi_transfer(device, (char*)WriteBuffer, NULL, 4);
}

void FixedGen(u8 channels, u16 fixedvalue) {
    // 0x55 - this is the dummy data for 12-bit dac
    WriteBuffer[3] = 0x55;
    // least significant 8-bits data
    WriteBuffer[2] = fixedvalue & 0xff;
    // channel number (0) | most significant 4-bits data
    WriteBuffer[1] = (channels << 4) | (fixedvalue >> 8);
    // 4 most significant bits don't care | write and update DAC command
    WriteBuffer[0] = 0x03;

    spi_transfer(device, (char*)WriteBuffer, NULL, 4);
}


void TriangleWaveGen(u8 channels, u8 numofcycles, u16 delay) {
    int i, j;

    WriteBuffer[3] = 0x55;
    WriteBuffer[0] = 0x03;

    if(numofcycles==0){
        for(;;){
            for(j=0; j< 4096; j++) {
                WriteBuffer[2] = j & 0xff;
                WriteBuffer[1] = (channels << 4) | ((j >> 8 ) & 0x0f);
                spi_transfer(device, (char*)WriteBuffer, NULL, 4);
                if(delay)
                    delay_us(delay);
            }
            for(j=4095; j>=0; j--) {
                WriteBuffer[2] = j & 0xff;
                WriteBuffer[1] = (channels << 4) | ((j >> 8 ) & 0x0f);
                spi_transfer(device, (char*)WriteBuffer, NULL, 4);
                if(delay)
                    delay_us(delay);
            }
        }
    }
    else {
        for(i=0; i<numofcycles ; i++) {
            for(j=0; j< 4096; j++) {
                WriteBuffer[2] = j & 0xff;
                WriteBuffer[1] = (channels << 4) | ((j >> 8 ) & 0x0f);
                spi_transfer(device, (char*)WriteBuffer, NULL, 4);
                if(delay)
                    delay_us(delay);
            }
            for(j=4095; j>=0; j--) {
                WriteBuffer[2] = j & 0xff;
                WriteBuffer[1] = (channels << 4) | ((j >> 8 ) & 0x0f);
                spi_transfer(device, (char*)WriteBuffer, NULL, 4);
                if(delay)
                    delay_us(delay);
            }
        }
    }
}



void ArbitraryWaveGen (u8 channels, u8 lengthOfMailbox) {
    int i, j;

    WriteBuffer[3] = 0x55;
    WriteBuffer[0] = 0x03;

    // read in from mailbox and write to pmod spi
    for(;;){
        for(j = 0; j < lengthOfMailbox; j++) {
            int num = (u32 *)MAILBOX_DATA(j);

			int val1 = (num >> 12) & 0x0fff;
		    int val2 = num & 0x0fff;

            WriteBuffer[2] = val1 & 0xff;
            WriteBuffer[1] = (channels << 4) | ((val1 >> 8 ) & 0x0f);
            spi_transfer(device, (char*)WriteBuffer, NULL, 4);

            WriteBuffer[2] = val2 & 0xff;
            WriteBuffer[1] = (channels << 4) | ((val2 >> 8 ) & 0x0f);
            spi_transfer(device, (char*)WriteBuffer, NULL, 4);
        }
    }
}


void dac(u8 channels, u8 mode, u16 fixedvalue, u8 numofcycles, u16 delay)
{

    switch(mode) {
        case 1:FixedGen(channels, fixedvalue); break;
        case 2:TriangleWaveGen(channels, numofcycles, delay); break;
        case 3:ArbitraryWaveGen(channels, numofcycles); break;
    }
}

int main(void)
{
    u16 delay;
    u32 cmd;
    u8 numofcycles;
    u16 fixedvalue;
    u8 mode, channels;

    device = spi_open(3, 2, 1, 0);
    device = spi_configure(device, 0, 1);

    RefOn();
    while(1){
        while((MAILBOX_CMD_ADDR & 0x01)==0);
        cmd = MAILBOX_CMD_ADDR;
        mode=0;
        if((cmd >> 1) & 0x01)
            mode=1;
        else if((cmd >> 2) & 0x01)
            mode=2;
        else if((cmd >> 3) & 0x01)
            mode=3;

        channels = (cmd >> 16) & 0x0f;
        fixedvalue = (cmd >> 20) & 0x0fff;

        if(((cmd >> 20) & 0x0fff)==0)
            // set to 1 second if the field is set to 0
            delay = 1000;
        else
            // multiple of approximate milliseconds
            delay=(cmd >> 20) & 0xfff;
        if(((cmd >> 8) & 0x000ff)==0)
            // indicate infinite number of samples
            numofcycles = 0;
        else
            // set to number of samples
            numofcycles=(cmd >> 8) & 0x0ff;

        dac(channels,mode,fixedvalue,numofcycles,delay);
        MAILBOX_CMD_ADDR = 0x0;
    }
    return 0;
}


